### DISCUSSION TITLE
Discovery Server not working over LAN

### DISCUSSION THREAD

**Discussion by maruniluk** [eProsima/Fast-DDS#2809] [Q&A]
Hello,
I am trying to work with Discovery Server. I tried and tested the Hello World example described here https://eprosima-discovery-server.readthedocs.io/en/latest/cpp_examples/HelloWorldExample.html. I also tried modifying my application according to this https://fast-dds.docs.eprosima.com/en/v2.7.0/fastdds/discovery/discovery_server.html#full-example. Unfortunately, I keep failing to make it work over LAN. It works as long as I use it on a single host. But it won't go through when I try it on a LAN. I tried it on several networks. Computers could ping each other in all situations, and regular discovery (using broadcasts) worked without problems. I tried both Windows and Linux versions. 
Do you have any ideas what might be wrong?

**Comment by EduPonz** [comment]
Hi @maruniluk ,

I'm afraid we are going to need more information. What is it exactly that you are doing? Can you post some minimal application? How are you instantiating the Server? Are you doing it within you code, or are you using [Fast DDS' CLI](https://fast-dds.docs.eprosima.com/en/v2.7.0/fastddscli/cli/cli.html#discovery)? How are you configuring the Clients? Is it via code, XML, or environment variable?

**Comment by maruniluk** [comment]
Thank you for your response. 
I tried running it using my code (snippets follows) and with provided example. Since the behaviour was still the same (single host works, network doesn't), my subsequent experiments boiled down to running the HelloWorld example from [here](https://eprosima-discovery-server.readthedocs.io/en/latest/cpp_examples/HelloWorldExample.html). I totally missed the Fast DDS CLI, will give it a try.

I have two computers with IPs `192.168.71.131` and `192.168.71.132`. Let's say I run the server and publisher on `192.168.71.131` and the subscriber on `192.168.71.132`. My commands would be this:

PC1 (192.168.71.131):
```
../../../../../local_setup.bash
./HelloWorldExampleDS publisher --ip=192.168.71.131:12345
```

```
../../../../../local_setup.bash
./HelloWorldExampleDS server --ip=0.0.0.0:12345
```

PC2 (192.168.71.132):
```
../../../../../local_setup.bash
./HelloWorldExampleDS subscriber --ip=192.168.71.131:12345
```
In all cases, I tried using default ports and some custom ones. I also tried providing the IP address to the server directly, but again, there was no change in behaviour.

When running it from code, I used examples from the documentation.
Server:
 ```
// Get default participant QoS
DomainParticipantQos server_qos = PARTICIPANT_QOS_DEFAULT;

// Set participant as SERVER
server_qos.wire_protocol().builtin.discovery_config.discoveryProtocol = eprosima::fastrtps::rtps::DiscoveryProtocol_t::SERVER;

// Set SERVER's GUID prefix
std::istringstream("44.53.00.5f.45.50.52.4f.53.49.4d.41") >> server_qos.wire_protocol().prefix;

// Set SERVER's listening locator for PDP
eprosima::fastrtps::rtps::Locator_t locator;
eprosima::fastrtps::rtps::IPLocator::setIPv4(locator, 192, 168, 71, 131);
locator.port = 11811;
server_qos.wire_protocol().builtin.metatrafficUnicastLocatorList.push_back(locator);

// Create SERVER
server_qos.name("dummy SERVER");
DomainParticipant* server = DomainParticipantFactory::get_instance()->create_participant(0, server_qos);
```

Client (in this case Subscriber):
```
// Get default participant QoS
DomainParticipantQos client_qos = PARTICIPANT_QOS_DEFAULT;

// Set participant as CLIENT
client_qos.wire_protocol().builtin.discovery_config.discoveryProtocol = eprosima::fastrtps::rtps::DiscoveryProtocol_t::CLIENT;

// Set SERVER's GUID prefix
eprosima::fastrtps::rtps::RemoteServerAttributes remote_server_att;
remote_server_att.ReadguidPrefix("44.53.00.5f.45.50.52.4f.53.49.4d.41");

// Set SERVER's listening locator for PDP
eprosima::fastrtps::rtps::Locator_t locator;
eprosima::fastrtps::rtps::IPLocator::setIPv4(locator, 192, 168, 71, 131);
locator.port = 11811;
remote_server_att.metatrafficUnicastLocatorList.push_back(locator);

// Add remote SERVER to CLIENT's list of SERVERs
client_qos.wire_protocol().builtin.discovery_config.m_DiscoveryServers.push_back(remote_server_att);

// Set ping period to 250 ms
client_qos.wire_protocol().builtin.discovery_config.discoveryServer_client_syncperiod = eprosima::fastrtps::Duration_t(0, 250000000);

DomainParticipantQos participantQos;
client_qos.name("dummy SUBSCRIBER");
participant_ = DomainParticipantFactory::get_instance()->create_participant(0, client_qos);
type_.register_type(participant_);
topic_ = participant_->create_topic("Topic", "DataType", TOPIC_QOS_DEFAULT);
subscriber_ = participant_->create_subscriber(SUBSCRIBER_QOS_DEFAULT, nullptr);
reader_ = subscriber_->create_datareader(topic_, DATAREADER_QOS_DEFAULT, &listener_);
```

**Comment by EduPonz** [comment]
Hi @maruniluk,

Since the question is really about how to set up discovery server, I'll let your example aside and focus on a much simple one, which is the one presented on the [Fast DDS Getting Started section](https://fast-dds.docs.eprosima.com/en/latest/fastdds/getting_started/getting_started.html) (the source code for it can be found [here](https://github.com/eProsima/Fast-DDS-docs/tree/master/code/Examples/C%2B%2B/DDSHelloWorld).

To that example, I have applied the following patch (which is simply adding Discovery Server config, mind to adjust your IP):

```diff
diff --git a/code/Examples/C++/DDSHelloWorld/src/HelloWorldPublisher.cpp b/code/Examples/C++/DDSHelloWorld/src/HelloWorldPublisher.cpp
index 909620cd..6d2be291 100644
--- a/code/Examples/C++/DDSHelloWorld/src/HelloWorldPublisher.cpp
+++ b/code/Examples/C++/DDSHelloWorld/src/HelloWorldPublisher.cpp
@@ -117,6 +117,17 @@ public:
         hello_.message("HelloWorld");
 
         DomainParticipantQos participantQos;
+        participantQos.wire_protocol().builtin.discovery_config.discoveryProtocol = eprosima::fastrtps::rtps::DiscoveryProtocol_t::SERVER;
+
+        // Set SERVER's GUID prefix
+        std::istringstream("44.53.00.5f.45.50.52.4f.53.49.4d.41") >> participantQos.wire_protocol().prefix;
+
+        // Set SERVER's listening locator for PDP
+        eprosima::fastrtps::rtps::Locator_t locator;
+        eprosima::fastrtps::rtps::IPLocator::setIPv4(locator, 192, 168, 1, 42);
+        locator.port = 11811;
+        participantQos.wire_protocol().builtin.metatrafficUnicastLocatorList.push_back(locator);
+
         participantQos.name("Participant_publisher");
         participant_ = DomainParticipantFactory::get_instance()->create_participant(0, participantQos);
 
diff --git a/code/Examples/C++/DDSHelloWorld/src/HelloWorldSubscriber.cpp b/code/Examples/C++/DDSHelloWorld/src/HelloWorldSubscriber.cpp
index c8d6f248..aac966d8 100644
--- a/code/Examples/C++/DDSHelloWorld/src/HelloWorldSubscriber.cpp
+++ b/code/Examples/C++/DDSHelloWorld/src/HelloWorldSubscriber.cpp
@@ -129,6 +129,15 @@ public:
     bool init()
     {
         DomainParticipantQos participantQos;
+        participantQos.wire_protocol().builtin.discovery_config.discoveryProtocol = eprosima::fastrtps::rtps::DiscoveryProtocol_t::CLIENT;
+        eprosima::fastrtps::rtps::RemoteServerAttributes remote_server_att;
+        remote_server_att.ReadguidPrefix("44.53.00.5f.45.50.52.4f.53.49.4d.41");
+        eprosima::fastrtps::rtps::Locator_t locator;
+        eprosima::fastrtps::rtps::IPLocator::setIPv4(locator, 192, 168, 1, 42);
+        locator.port = 11811;
+        remote_server_att.metatrafficUnicastLocatorList.push_back(locator);
+        participantQos.wire_protocol().builtin.discovery_config.m_DiscoveryServers.push_back(remote_server_att);
+        participantQos.wire_protocol().builtin.discovery_config.discoveryServer_client_syncperiod = eprosima::fastrtps::Duration_t(0, 250000000);
         participantQos.name("Participant_subscriber");
         participant_ = DomainParticipantFactory::get_instance()->create_participant(0, participantQos);
```

This works when running the publisher in one machine and the subscriber in another. Please do let me know how it goes.

**Reply by maruniluk** [reply]
Yes, it worked!
Do know what was I doing wrong? I understand I could have messed up the code setup, but I certainly didn't expect the [HelloWorld](https://eprosima-discovery-server.readthedocs.io/en/latest/cpp_examples/HelloWorldExample.html) example to fail.
